### PR TITLE
Add drop-connected in efficientnet & refine the bifpn

### PR DESCRIPTION
**Pull request by zimoqingfeng**
The map about the refined EfficientDet-D0 could achieve 0.341 in coco minval
image

### CODE REVIEW

File: configs/efficientdet_d0.yml [hunk 1]
```diff
@@ -39,9 +37,9 @@ EfficientHead:
   alpha: 0.25
   delta: 0.1
   output_decoder:
-    score_thresh: 0.05   # originally 0.
+    score_thresh: 0.0
     nms_thresh: 0.5
-    pre_nms_top_n: 1000  # originally 5000
+    pre_nms_top_n: 5000
```
Comment (willthefrog) [inline configs/efficientdet_d0.yml:42]: this is per class top n, 5000 seems like overkill

File: ppdet/modeling/backbones/bifpn.py [hunk 2]
```diff
@@ -41,7 +41,8 @@ def __call__(self, inputs, name=''):
             groups=self.num_chan,
             param_attr=ParamAttr(
                 initializer=Xavier(), name=name + '_dw_w'),
-            bias_attr=False)
+            bias_attr=False,
+            use_cudnn=False)
```
Comment (willthefrog) [inline ppdet/modeling/backbones/bifpn.py:45]: according to previous benchmark, on cudnn 7.6+ depthwise is indeed faster, especially during inference

File: ppdet/modeling/backbones/bifpn.py [hunk 3]
```diff
@@ -66,81 +67,114 @@ def __call__(self, inputs, name=''):
...
         bigates = fluid.layers.relu(self.bigates)
-        trigates /= fluid.layers.reduce_sum(
-            trigates, dim=1, keep_dim=True) + self.eps
-        bigates /= fluid.layers.reduce_sum(
-            bigates, dim=1, keep_dim=True) + self.eps
+        trigates /= fluid.layers.reduce_sum(trigates, dim=1, keep_dim=True) + self.eps
+        bigates /= fluid.layers.reduce_sum(bigates, dim=1, keep_dim=True) + self.eps
 
-        feature_maps = list(inputs)  # make a copy
+        feature_maps = list(inputs)  # make a copy, 依次是 [P3, P4, P5, P6, P7]
```
Comment (willthefrog) [inline ppdet/modeling/backbones/bifpn.py:143]: avoid non ASCII chars in source code

**Review by willthefrog — CHANGES_REQUESTED** [review]
could use some simplification, also, maybe document architecture changes in PR description for future reference

File: ppdet/modeling/backbones/bifpn.py [hunk 4]
```diff
@@ -164,39 +183,80 @@ def __init__(self, num_chan, repeat=3, levels=5):
...
+                bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                name='resample2_bn_{}'.format(idx))
+            p4_2_p5_2.append(feat)
 
-        biFPN = BiFPNCell(self.num_chan, self.levels)
+        biFPN = BiFPNCell(self.num_chan, self.levels, len(inputs))
         for r in range(self.repeat):
-            feats = biFPN(feats, 'bifpn_{}'.format(r))
+            if r == 0:
+                feats = biFPN(feats, cell_name='bifpn_{}'.format(r), is_first_time=True, p4_2_p5_2=p4_2_p5_2)
```
Comment (willthefrog) [inline ppdet/modeling/backbones/bifpn.py:258]: just handle p4 and p5 when populate the `feats` list, so that there is no need for `is_first_time` and `p4_2_p5_2`?

File: configs/efficientdet_d0.yml [hunk 5]
```diff
@@ -39,7 +37,7 @@ EfficientHead:
   alpha: 0.25
   delta: 0.1
   output_decoder:
-    score_thresh: 0.05   # originally 0.
+    score_thresh: 0.0
```
Comment (willthefrog) [inline configs/efficientdet_d0.yml:40]: would like some ablation experiments, what is the mAP like when `sync_bn` and `score_thresh` is unchanged?

File: ppdet/modeling/backbones/efficientnet.py [hunk 6]
```diff
@@ -88,37 +96,44 @@ def get_model_params(scale):
...
+        'l2': (4.3, 5.3),
     }
 
     w, d = params_dict[scale]
 
     global_params = GlobalParams(
+        blocks_args=block_strings,
         batch_norm_momentum=0.99,
         batch_norm_epsilon=1e-3,
+        drop_connect_rate=0 if scale == 'b0' else 0.2,
```
Comment (willthefrog) [inline ppdet/modeling/backbones/efficientnet.py:108]: would make more sense to split the drop connect part as a separate PR, as it is actually  unused for d0.

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: configs/efficientdet_d0.yml (hunk 5 above)
Comment (zimoqingfeng) [reply configs/efficientdet_d0.yml:40]: Ablation experiments have been done. Without the lower score_thresh and the sync_bn, the mAP could only achieve 0.335 on coco minival.
image

File: ppdet/modeling/backbones/bifpn.py (hunk 4 above)
Comment (zimoqingfeng) [reply ppdet/modeling/backbones/bifpn.py:258]: OK, and I would like to point out that it is different when the BIFPN has been used at the first time. Another 3rd party implementation could be referenced. https://github.com/bubbliiiing/efficientdet-pytorch/blob/7a89bf2b2f6b482b4d224f9c54a275e911234b71/nets/efficientdet.py#L153

Comment (willthefrog) [reply ppdet/modeling/backbones/bifpn.py]: again, why can't be done when populate the initial `feats` list , before passing into the first bifpn cell?

File: configs/efficientdet_d0.yml (hunk 5 above)
Comment (willthefrog) [reply configs/efficientdet_d0.yml:40]: so basically all the gain in mAP is from these 2 configuration change?

File: ppdet/modeling/backbones/bifpn.py (hunk 4 above)
Comment (zimoqingfeng) [reply ppdet/modeling/backbones/bifpn.py]: i have modified it in the third commit, have you seen?

File: configs/efficientdet_d0.yml (hunk 5 above)
Comment (zimoqingfeng) [reply configs/efficientdet_d0.yml:40]: based on the modified bifpn

Comment (zimoqingfeng) [reply configs/efficientdet_d0.yml:40]: I have to point out **again**, ur bifpn is not aligned with the official tensorflow implementation, as well as the drop connected in efficientnet backbone. I think it will be helpful for EfficientDet-D1 to D7x. If ur team confirm it is useless, I think you just hit that I could withdraw the submission.

Comment (willthefrog) [reply configs/efficientdet_d0.yml:40]: my point is, neither of these settings change are free,  will increase either training time or inference latency, and with the bifpn change alone (which also adds additional cost, albeit small) , mAP actually dropped 0.3, am I understanding it correctly?

File: ppdet/modeling/backbones/bifpn.py [hunk 7]
```diff
@@ -164,39 +183,76 @@ def __init__(self, num_chan, repeat=3, levels=5):
...
+                    epsilon=1e-04,
+                    param_attr=ParamAttr(
+                        initializer=Constant(1.0), regularizer=L2Decay(0.)),
+                    bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                    name='resample_bn_{}'.format(idx))
+            else:
+                feat = inputs[idx]
+            feats.append(feat)
+        # Build additional input features that are not from backbone.
+        # P_7 layer we just use pool2d without conv layer & bn, for the same channel with P_6.
```
Comment (willthefrog) [inline ppdet/modeling/backbones/bifpn.py:208]: could do without the code reference.

**Comment by yzl19940819** [comment]
@zimoqingfeng @willthefrog Shall we solve the ci problem and continue merge or close the request?       PM  yuzhiliang

**Comment by yzl19940819** [comment]
@zimoqingfeng 嗨，我看目前事CI测试出现了问题？另外根据上面的内容，能否在辛苦做一个测试，即保持结构不变的情况下，改变超参数看看实验结果呢？

**Comment by zimoqingfeng** [comment]
image

The results which only with the sync_bn (without modifying the bifpn).

**Comment by willthefrog** [comment]
> The results which only with the sync_bn (without modifying the bifpn).

what about the nms settings?

**Comment by zimoqingfeng** [comment]
Based on the sync_bn and the modified nms settings:
image

Based on the sync_bn & the modified nms settings & the lower threshold:
image

**Comment by willthefrog** [comment]
Please fix the CI errors (formatting, etc)
